File: samples/MessageCenterSample/MessageCenterSample-ObjC/ViewController.m
#import "ViewController.h"
#import "SwrveSDK.h"
#import "SwrveCampaign.h"

@implementation ViewController

@synthesize campaigns;

- (void)viewDidLoad {
    [super viewDidLoad];
    [self refreshDataSource];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self refreshDataSource];
    [self.tableView reloadData];

    // Observe for the new campaigns
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(reloadDataAndView) name:@"SwrveUserResourcesUpdated" object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(reloadDataAndView) name:@"SwrveMessageWillBeHidden" object:nil];
}

- (void) reloadDataAndView
{
    [self refreshDataSource];
    [self.tableView reloadData];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    // Remove observe
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)refreshDataSource {
    // Obtain latest campaigns and order by the campaign dateStart property
    NSSortDescriptor *dateDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"dateStart" ascending:NO];
    NSArray *sortDescriptors = [NSArray arrayWithObject:dateDescriptor];
    self.campaigns = [[SwrveSDK messageCenterCampaigns] sortedArrayUsingDescriptors:sortDescriptors];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.campaigns.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"MessageCenterCell"];

    // Configure the cell
    SwrveCampaign *campaign = [self.campaigns objectAtIndex:indexPath.row];

    // Campaign subject
    cell.textLabel.text = campaign.subject;
    [cell.textLabel setBackgroundColor:[UIColor clearColor]];
    [cell.detailTextLabel setBackgroundColor:[UIColor clearColor]];

    // Format the start date for display
    NSDateFormatter *dformat = [NSDateFormatter new];
    [dformat setDateFormat:@"MMMM dd, yyyy (EEEE) HH:mm:ss z Z"];

    // Campaign start date, change cell background colour based on seen / unseen status
    switch(campaign.state.status) {
        case SWRVE_CAMPAIGN_STATUS_UNSEEN:
            cell.detailTextLabel.text = [dformat stringFromDate:campaign.dateStart];
            [cell setBackgroundColor:[UIColor colorWithRed:1.0f green:0.0f blue:0.0f alpha:0.4f]];

            break;
        case SWRVE_CAMPAIGN_STATUS_SEEN:
            cell.detailTextLabel.text = [dformat stringFromDate:campaign.dateStart];
            [cell setBackgroundColor:[UIColor colorWithRed:0.0f green:1.0f blue:0.0f alpha:0.4f]];
            break;
        default:
            cell.detailTextLabel.text = @"Other";
            break;
    }

    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 80.0;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    // Display the campaign when clicked
    [SwrveSDK showMessageCenterCampaign:[self.campaigns objectAtIndex:indexPath.row]];
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath {
    return UITableViewCellEditingStyleDelete;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Remove campaign from the sliding button
        [tableView beginUpdates];
        [SwrveSDK removeMessageCenterCampaign:[self.campaigns objectAtIndex:indexPath.row]];
        [self refreshDataSource];
        [tableView deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationLeft];
        [tableView endUpdates];
    }
}

@end

File: samples/MessageCenterSample/MessageCenterSample-Swift/ViewController.swift
import SwrveSDK
import UIKit

class ViewController: UITableViewController {

    var campaigns: NSArray!

    override func viewDidLoad() {
        super.viewDidLoad()
        refreshDataSource()
        self.tableView.reloadData()

        // Observe for the new campaigns
        NotificationCenter.default.addObserver(
            self, selector: #selector(ViewController.reloadData(_:)), name: NSNotification.Name(rawValue: "SwrveUserResourcesUpdated"), object: nil)
        NotificationCenter.default.addObserver(
            self, selector: #selector(ViewController.reloadData(_:)), name: NSNotification.Name(rawValue: "SwrveMessageWillBeHidden"), object: nil)
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        NotificationCenter.default.removeObserver(self)
    }

    func refreshDataSource() {
        let unsortedCampaigns: NSArray = SwrveSDK.messageCenterCampaigns() as NSArray
        let descriptor: NSSortDescriptor = NSSortDescriptor(key: "dateStart", ascending: false)
        campaigns = unsortedCampaigns.sortedArray(using: [descriptor]) as NSArray
    }

    @objc func reloadData(_ notification: Notification) {
        refreshDataSource()
        self.tableView.reloadData()
    }
}

// MARK: UITableViewDelegate and UITableViewDataSource extension
extension ViewController {

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let campaign = campaigns[indexPath.row] as? SwrveCampaign

        if let campaign = campaign {
            _ = SwrveSDK.showMessageCenter(campaign)
        }
    }

    override func tableView(_ tableView: UITableView, canMoveRowAt indexPath: IndexPath) -> Bool {
        true
    }

    override func tableView(_ tableView: UITableView, editingStyleForRowAt indexPath: IndexPath) -> UITableViewCell.EditingStyle {
        UITableViewCell.EditingStyle.delete
    }

    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle == UITableViewCell.EditingStyle.delete {
            guard let campaign = campaigns[indexPath.row] as? SwrveCampaign else { return }
            tableView.beginUpdates()
            SwrveSDK.removeMessageCenter(campaign)
            tableView.deleteRows(at: [indexPath], with: UITableView.RowAnimation.left)
            tableView.endUpdates()
        }
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "MessageCenterCell", for: indexPath)
        let baseCampaign = campaigns[indexPath.row] as? SwrveCampaign
        cell.textLabel?.text = baseCampaign?.messageCenterDetails?.subject ?? ""

        let dformat: DateFormatter = DateFormatter()
        dformat.dateFormat = "MMMM dd, yyyy (EEEE) HH:mm:ss z Z"
        cell.detailTextLabel?.text = dformat.string(from: (baseCampaign?.dateStart)!)

        // Campaign cell background colour based on seen / unseen status
        if let state = baseCampaign?.state.status {
            switch state {
            case .unseen:
                cell.backgroundColor = UIColor.init(red: 1, green: 0, blue: 0, alpha: 0.4)
            case .seen:
                cell.backgroundColor = UIColor.init(red: 0, green: 1, blue: 0, alpha: 0.4)
            default:
                break
            }
        }
        return cell
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        campaigns.count
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        1
    }
}
